crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+use std::io::{Cursor, Read};
+use std::sync::Arc;
+
+use itertools::Itertools;
+use zip::{read::ZipFile, result::ZipResult, ZipArchive};
+
+use crate::file_system::{FileRevision, FileSystem, FileSystemPath, FileType, Metadata, Result};
+
+#[derive(Debug, Clone)]
+pub struct VendoredFileSystem {
+    inner: VendoredZipArchive,
+}
+
+impl VendoredFileSystem {
+    const READ_ONLY_PERMISSIONS: u32 = 0o444;
+
+    pub fn new(raw_bytes: Arc<[u8]>) -> Result<Self> {
```

Comment:
Can we use a `&'static [u8]` instead of an `Arc<[u8]>`
```suggestion
    pub fn new(raw_bytes: &'static [u8]) -> Result<Self> {
```

Reply:
I had that initially, but it made it much harder to test it using a mock zipfile

Reply:
This relates to the conversation we're having at https://github.com/astral-sh/ruff/pull/11863#discussion_r1639342264. If we just committed a binary zip to the repo and used that in the test, then we could easily use `&'static` for the implementation. But IMHO that would be a really bad test, because we wouldn't be able to easily see what we'd put in the zip archive, so I feel like any assertions we'd make would honestly not have that much value

Reply:
That's fair. I think we should allow both then. Using `&'static [u8]` has the benefit that we can pass the bytes directly from the binary without having to copy the zip to the heap first (it's already on the heap because the binary is on the heap, but we can avoid duplicating it). 

You can implement it by having a 

```rust
enum ZipData {
	Static(&'static [u8]),
	Dynamic(Arc<[u8]) // Or Box if we can drop the `Clone` requirement
}
```

And then provide a  `new_static` and `new` function on the `VendoredFilesystem`

Reply:
Brilliant, thank you!